libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -97,321 +97,412 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
+        _attitude_target_ang_vel.z = input_shaping_ang_vel(_attitude_target_ang_vel.z, yaw_rate_rads, get_accel_yaw_max_radss());
     } else {
-        _att_target_ang_vel_rads.z = yaw_rate_bf_rads;
+        // When axis acceleration limiting is disabled, the attitude target is simply rotated using the input rate and the input rate
+        // is fed forward into the rate controller.
+        _attitude_target_ang_vel.x = roll_rate_rads;
+        _attitude_target_ang_vel.y = pitch_rate_rads;
+        _attitude_target_ang_vel.z = yaw_rate_rads;
     }
 
-    // Compute quaternion target attitude
-    Quaternion att_target_quat;
-    att_target_quat.from_euler(_att_target_euler_rad.x,_att_target_euler_rad.y,_att_target_euler_rad.z);
+    // Call quaternion attitude controller
+    attitude_controller_run_quat();
+}
 
-    // Rotate quaternion target attitude using computed rate
-    att_target_quat.rotate(_att_target_ang_vel_rads*_dt);
-    att_target_quat.normalize();
+// Calculates the body frame angular velocities to follow the target attitude
+void AC_AttitudeControl::attitude_controller_run_quat()
+{
+    // Retrieve quaternion vehicle attitude
+    // TODO add _ahrs.get_quaternion()
+    Quaternion attitude_vehicle_quat;
+    attitude_vehicle_quat.from_rotation_matrix(_ahrs.get_rotation_body_to_ned());
 
-    // Call attitude controller
-    attitude_controller_run_quat(att_target_quat, _att_target_ang_vel_rads);
+    // Compute attitude error
+    Vector3f attitude_error_vector;
+    thrust_heading_rotation_angles(_attitude_target_quat, attitude_vehicle_quat, attitude_error_vector, _thrust_error_angle);
 
-    // Keep euler derivative updated
-    ang_vel_to_euler_rate(Vector3f(_ahrs.roll,_ahrs.pitch,_ahrs.yaw), _ang_vel_target_rads, _att_target_euler_rate_rads);
+    // Compute the angular velocity target from the attitude error
+    _rate_target_ang_vel = update_ang_vel_target_from_att_error(attitude_error_vector);
+
+    // Add feedforward term that attempts to ensure that roll and pitch errors rotate with the body frame rather than the reference frame.
+    _rate_target_ang_vel.x += attitude_error_vector.y * _ahrs.get_gyro().z;
+    _rate_target_ang_vel.y += -attitude_error_vector.x * _ahrs.get_gyro().z;
+
+    // Add the angular velocity feedforward, rotated into vehicle frame
+    Quaternion attitude_target_ang_vel_quat = Quaternion(0.0f, _attitude_target_ang_vel.x, _attitude_target_ang_vel.y, _attitude_target_ang_vel.z);
+    Quaternion attitude_error_quat = attitude_vehicle_quat.inverse() * _attitude_target_quat;
+    Quaternion target_ang_vel_quat = attitude_error_quat.inverse()*attitude_target_ang_vel_quat*attitude_error_quat;
+
+    // Correct the thrust vector and smoothly add feedforward and yaw input
+    if(_thrust_error_angle > AC_ATTITUDE_THRUST_ERROR_ANGLE*2){
+        _rate_target_ang_vel.z = _ahrs.get_gyro().z;
+    }else if(_thrust_error_angle > AC_ATTITUDE_THRUST_ERROR_ANGLE){
```

Comment:
Sorry for my stupidity, can you tell me what the correct formating would look like? (or are you referring to 2 instead of 2.0f)